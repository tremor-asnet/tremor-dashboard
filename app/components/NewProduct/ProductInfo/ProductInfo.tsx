"use client";

import dynamic from "next/dynamic";

// Libs
import { Controller, useFormContext } from "react-hook-form";

// Components
import {
  Text,
  Flex,
  Card,
  TextInput,
  Select,
  SelectItem,
  NumberInput,
} from "@tremor/react";
const QuillEditor = dynamic(() => import("react-quill"), { ssr: false });

// Constants
import {
  MESSAGES_ERROR,
  CATEGORY_PRODUCT,
  NUMBER_REGEX,
  DECIMAL_REGEX,
} from "@/constants";

// Types
import { SelectOptionData } from "@/types";

// Styles
import "@/styles/form.css";
import "react-quill/dist/quill.snow.css"; // Import Quill styles

const ProductInfo = () => {
  const {
    control,
    formState: { errors },
  } = useFormContext();
  const { name } = errors || {};
  const nameErrorMessage = name?.message?.toString() || "";

  return (
    <Card className="w-full dark:bg-dark-tremor-primary rounded-lg shadow-box-icon-default ring-0">
      <Text className="text-primary dark:text-white font-bold text-xl mb-8">
        Product Information
      </Text>

      <Flex className="items-start gap-5">
        <Flex flexDirection="col" className="gap-4">
          <Controller
            control={control}
            rules={{
              required: MESSAGES_ERROR.NAME_REQUIRED,
              minLength: {
                value: 4,
                message: MESSAGES_ERROR.NAME_MIN_LENGTH,
              },
            }}
            render={({ field }) => (
              <div className="w-full">
                <Text className="text-secondary dark:text-lighter mb-2">
                  Name
                </Text>
                <TextInput
                  id="name"
                  placeholder="Name"
                  className="py-1 w-full dark:text-white hover:bg-transparent bg-transparent dark:bg-transparent focus:bg-transparent rounded-b-none border-l-0 border-r-0 border-t-0 border-b-1 focus:border-b-2 focus:outline-none focus:border-tremor-brand-subtle dark:border-light dark:focus:border-white shadow-none hover:bg-transparent ring-0"
                  autoFocus
                  required
                  {...field}
                />
                {nameErrorMessage && (
                  <p className="pt-1 text-[11px] xs:text-xs text-red-500">
                    {nameErrorMessage}
                  </p>
                )}
              </div>
            )}
            name="productName"
          />

          <Controller
            control={control}
            render={({ field }) => (
              <div className="w-full">
                <Text className="text-secondary dark:text-lighter mb-2">
                  Description <span className="text-xs">(optional)</span>
                </Text>
                <QuillEditor
                  placeholder="Content goes here..."
                  className="w-full rounded-md text-secondary dark:text-white"
                  {...field}
                />
              </div>
            )}
            name="description"
          />
        </Flex>
        <Flex flexDirection="col">
          <Controller
            control={control}
            rules={{
              pattern: { value: DECIMAL_REGEX, message: "Invalid weight" },
            }}
            render={({ field }) => (
              <div className="w-full mb-4">
                <Text className="text-secondary dark:text-lighter mb-2">
                  Weight
                </Text>
                <TextInput
                  id="weight"
                  placeholder="Weight"
                  className="py-1 w-full dark:text-white hover:bg-transparent bg-transparent dark:bg-transparent focus:bg-transparent rounded-b-none border-l-0 border-r-0 border-t-0 border-b-1 focus:border-b-2 focus:outline-none focus:border-tremor-brand-subtle dark:border-light dark:focus:border-white shadow-none hover:bg-transparent ring-0"
                  {...field}
                />
              </div>
            )}
            name="weight"
          />

          <Controller
            control={control}
            rules={{
              pattern: {
                value: NUMBER_REGEX,
                message: "Invalid quantity number",
              },
            }}
            render={({ field: { onChange, value } }) => (
              <div className="w-full mb-4">
                <Text className="text-secondary dark:text-lighter mb-2">
                  Quantity
                </Text>
                <NumberInput
                  enableStepper={false}
                  onValueChange={onChange}
                  value={value}
                  className="py-1 w-full dark:text-white hover:bg-transparent bg-transparent dark:bg-transparent focus:bg-transparent rounded-b-none border-l-0 border-r-0 border-t-0 border-b-1 focus:border-b-2 focus:outline-none focus:border-tremor-brand-subtle dark:border-light dark:focus:border-white shadow-none hover:bg-transparent ring-0"
                />
              </div>
            )}
            name="quantity"
          />

          <Controller
            control={control}
            render={({ field: { value, onChange } }) => (
              <div className="w-full mb-4">
                <Text className="text-secondary mb-3 dark:text-lighter mb-2">
                  Category
                </Text>
                <Select
                  placeholder="Clothing"
                  className="select-custom dark:text-white dark:border-light dark:focus:border-white"
                  value={value.toString()}
                  onValueChange={onChange}>
                  {CATEGORY_PRODUCT.map((item: SelectOptionData) => (
                    <SelectItem key={item.value} value={item.value}>
                      {item.option}
                    </SelectItem>
                  ))}
                </Select>
              </div>
            )}
            name="category"
          />
        </Flex>
      </Flex>
    </Card>
  );
};

export default ProductInfo;
