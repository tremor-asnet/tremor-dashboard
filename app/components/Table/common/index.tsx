/*
Please write down export in alphabet order
*/

export * from "./CustomAvatarName";
export * from "./CustomCheckBoxField";
export * from "./CustomDateFormat";
export * from "./CustomList";
export * from "./CustomNumberFormat";
export * from "./CustomStatus";
export * from "./CustomQuantity";
