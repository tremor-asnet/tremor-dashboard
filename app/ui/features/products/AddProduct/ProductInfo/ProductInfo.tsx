"use client";

// Libs
import dynamic from "next/dynamic";
import { Control, Controller, FieldErrors } from "react-hook-form";
import { KeyboardEvent } from "react";

// Components
import { Text } from "@tremor/react";
import { InputField, SelectField } from "@/ui/components";

// Constants
import { CATEGORY_PRODUCT, MESSAGES_ERROR, EXCEPT_KEYS } from "@/constants";

// Types
import { NewInfo } from "@/types";

// Css
import "@/styles/form.css";
import "react-quill/dist/quill.snow.css";
import "@/styles/quill.css";

const QuillEditor = dynamic(() => import("react-quill"), {
  ssr: false,
  loading: () => (
    <div className="flex justify-center items-center h-44 w-full">
      <Text className="font-semibold text-secondary dark:text-white">
        Loading Quill Editor ...
      </Text>
    </div>
  ),
});

interface ProductInfoProps {
  control: Control<NewInfo>;
  errors: FieldErrors<NewInfo>;
}

const ProductInfo = ({ control, errors }: ProductInfoProps) => {
  const errorNameMsg = errors.productName?.message;

  const handleWeightKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    EXCEPT_KEYS.POSITIVE_DOUBLE.includes(e.key) && e.preventDefault();
  };

  const handleQuantityKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    ["e", "E", "+", "-", "."].includes(e.key) && e.preventDefault();
  };

  return (
    <div className="grid grid-cols-1 sm:grid-cols-2 gap-6 items-start">
      <Controller
        name="productName"
        control={control}
        rules={{
          required: MESSAGES_ERROR.FIELD_REQUIRED,
          minLength: {
            value: 4,
            message: MESSAGES_ERROR.MIN_LENGTH_4,
          },
        }}
        render={({ field }) => (
          <div>
            <InputField id="add-product-name" label="Name" {...field} />
            <Text className="pt-1 text-xs text-red-500 dark:text-red-500">
              {errorNameMsg}
            </Text>
          </div>
        )}
      />

      {/* Prevent input e, E, +, - to weight field */}
      <Controller
        name="weight"
        control={control}
        render={({ field }) => (
          <InputField
            id="add-product-weight"
            type="number"
            label="Weight"
            {...field}
            onKeyDown={handleWeightKeyDown}
          />
        )}
      />

      <Controller
        name="description"
        control={control}
        render={({ field }) => (
          <div className="w-full">
            <Text className="text-secondary text-sm dark:text-lighter">
              Description <span className="text-xs">(optional)</span>
            </Text>
            <QuillEditor
              theme="snow"
              placeholder="Some initial bold text"
              className="mt-3 dark:text-white"
              {...field}
            />
          </div>
        )}
      />

      <div className="flex flex-col gap-6">
        <div className="mb-2">
          <Controller
            name="category"
            control={control}
            render={({ field: { value, onChange } }) => {
              const convertedValue = value.toString();
              return (
                <SelectField
                  label="Category"
                  options={CATEGORY_PRODUCT}
                  value={convertedValue}
                  onChange={onChange}
                />
              );
            }}
          />
        </div>

        <div className="mb-2">
          <Controller
            name="quantity"
            control={control}
            render={({ field }) => (
              <InputField
                id="add-product-quantity"
                type="number"
                label="Quantity"
                onKeyDown={handleQuantityKeyDown}
                {...field}
              />
            )}
          />
        </div>

        <div className="mb-2">
          <Controller
            name="providerName"
            control={control}
            render={({ field }) => (
              <InputField
                id="add-product-provider"
                label="Provider Name"
                {...field}
              />
            )}
          />
        </div>
      </div>
    </div>
  );
};

export default ProductInfo;
