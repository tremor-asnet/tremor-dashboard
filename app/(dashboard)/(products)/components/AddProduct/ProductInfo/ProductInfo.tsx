"use client";

// Libs
import dynamic from "next/dynamic";
import { Control, Controller, FieldErrors } from "react-hook-form";
import { KeyboardEvent } from "react";

// Components
import { Text } from "@tremor/react";
import InputField from "@/components/common/CustomField/InputField/InputField";
import SelectField from "@/components/common/CustomField/SelectField/SelectField";

// Constants
import { CATEGORY_PRODUCT, MESSAGES_ERROR } from "@/constants";
import { EXCEPT_KEYS } from "@/constants/common";

// Types
import { NewInfo } from "@/types";

// Css
import "@/styles/form.css";
import "react-quill/dist/quill.snow.css";
import "@/styles/quill.css";

const QuillEditor = dynamic(() => import("react-quill"), {
  ssr: false,
  loading: () => (
    <div className="flex justify-center items-center h-44 w-full">
      <p className="text-sm font-semibold text-[#7b809a] dark:text-white">
        Loading Quill Editor ...
      </p>
    </div>
  ),
});

interface ProductInfoProps {
  control: Control<NewInfo>;
  errors: FieldErrors<NewInfo>;
}

const ProductInfo = ({ control, errors }: ProductInfoProps) => {
  const errorNameMsg = errors.productName?.message;

  const handleWeightKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    EXCEPT_KEYS.POSITIVE_DOUBLE.includes(e.key) && e.preventDefault();
  };

  const handleQuantityKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    ["e", "E", "+", "-", "."].includes(e.key) && e.preventDefault();
  };

  return (
    <div className="grid grid-cols-1 sm:grid-cols-2 gap-6 items-start">
      <Controller
        name="productName"
        control={control}
        rules={{
          required: MESSAGES_ERROR.FIELD_REQUIRED,
          minLength: {
            value: 4,
            message: MESSAGES_ERROR.MIN_LENGTH_4,
          },
        }}
        render={({ field }) => (
          <div>
            <InputField
              type="text"
              isLabelTransform={true}
              label="Name"
              {...field}
            />
            <p className="product-info-err-msg">{errorNameMsg}</p>
          </div>
        )}
      />

      {/* Prevent input e, E, +, - to weight field */}
      <Controller
        name="weight"
        control={control}
        render={({ field }) => (
          <InputField
            type="number"
            label="Weight"
            isLabelTransform={true}
            {...field}
            onKeyDown={handleWeightKeyDown}
          />
        )}
      />

      <Controller
        name="description"
        control={control}
        render={({ field }) => (
          <div className="w-full flex flex-col h-48">
            <Text className="text-secondary dark:text-lighter mb-2">
              Description <span className="text-xs">(optional)</span>
            </Text>
            <QuillEditor
              theme="snow"
              placeholder="Some initial bold text"
              className="mt-2 dark:text-white"
              {...field}
            />
          </div>
        )}
      />

      <div className="flex flex-col gap-6">
        <Controller
          name="category"
          control={control}
          render={({ field }) => (
            <div>
              <h3 className="text-tremor-default text-secondary dark:text-white">
                Category
              </h3>
              <SelectField options={CATEGORY_PRODUCT} {...field} />
            </div>
          )}
        />

        <Controller
          name="quantity"
          control={control}
          render={({ field }) => (
            <InputField
              type="number"
              isLabelTransform={true}
              label="Quantity"
              onKeyDown={handleQuantityKeyDown}
              {...field}
            />
          )}
        />

        <Controller
          name="providerName"
          control={control}
          render={({ field }) => (
            <InputField
              type="text"
              label="Provider Name"
              isLabelTransform={true}
              {...field}
            />
          )}
        />
      </div>
    </div>
  );
};

export default ProductInfo;
