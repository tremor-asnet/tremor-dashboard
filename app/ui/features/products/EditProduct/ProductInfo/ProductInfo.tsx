"use client";

import dynamic from "next/dynamic";
import { KeyboardEvent } from "react";

// Libs
import { Controller, useFormContext } from "react-hook-form";

// Components
import { Text, Flex, Card } from "@tremor/react";
import { InputField, SelectField } from "@/ui/components";
const QuillEditor = dynamic(() => import("react-quill"), { ssr: false });

// Constants
import { MESSAGES_ERROR, CATEGORY_PRODUCT, NAME_REGEX } from "@/constants";
import { EXCEPT_KEYS } from "@/constants/common";

// Styles
import "@/styles/form.css";
import "react-quill/dist/quill.snow.css"; // Import Quill styles

// Hooks
import useFocusFieldError from "@/hooks/useFocusFieldError";

const ProductInfo = () => {
  const { control, formState } = useFormContext();

  useFocusFieldError(formState);

  const handleOnKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    EXCEPT_KEYS.POSITIVE_DOUBLE.includes(e.key) && e.preventDefault();
  };

  return (
    <Card className="w-full dark:bg-dark-tremor-primary rounded-lg shadow-box-icon-default ring-0">
      <Text className="text-primary dark:text-white font-bold text-xl mb-8">
        Product Information
      </Text>

      <Flex
        flexDirection="col"
        alignItems="start"
        className="md:flex-row gap-6">
        <Flex flexDirection="col" className="gap-4">
          <Controller
            control={control}
            rules={{
              required: MESSAGES_ERROR.NAME_REQUIRED,
              pattern: {
                value: NAME_REGEX,
                message: MESSAGES_ERROR.INVALID_NAME,
              },
              minLength: {
                value: 4,
                message: MESSAGES_ERROR.NAME_MIN_LENGTH,
              },
            }}
            render={({ field, formState: { errors } }) => {
              const productNameErrorMessage = errors.productName?.message || "";

              return (
                <div className="w-full mb-4">
                  <InputField
                    id="edit-name"
                    label="Name"
                    errorMessage={productNameErrorMessage}
                    {...field}
                  />
                </div>
              );
            }}
            name="productName"
          />

          <Controller
            control={control}
            render={({ field }) => (
              <div className="w-full">
                <Text className="text-secondary text-sm dark:text-lighter">
                  Description <span className="text-xs">(optional)</span>
                </Text>
                <QuillEditor
                  placeholder="Content goes here..."
                  className="w-full rounded-md text-secondary dark:text-white pt-2.5"
                  {...field}
                />
              </div>
            )}
            name="description"
          />
        </Flex>
        <Flex flexDirection="col" className="gap-6">
          <Controller
            control={control}
            rules={{
              required: MESSAGES_ERROR.FIELD_REQUIRED,
              min: {
                value: 0,
                message: MESSAGES_ERROR.NEGATIVE_NUMBER,
              },
            }}
            render={({ field, formState: { errors } }) => {
              const weightErrorMessage = errors.weight?.message || "";

              return (
                <div className="w-full mb-1">
                  <InputField
                    id="edit-weight"
                    type="number"
                    label="Weight"
                    errorMessage={weightErrorMessage}
                    onKeyDown={handleOnKeyDown}
                    {...field}
                  />
                </div>
              );
            }}
            name="weight"
          />

          <Controller
            control={control}
            rules={{
              required: MESSAGES_ERROR.FIELD_REQUIRED,
              min: {
                value: 0,
                message: MESSAGES_ERROR.NEGATIVE_NUMBER,
              },
            }}
            render={({ field, formState: { errors } }) => {
              const quantityErrorMessage = errors.quantity?.message || "";

              return (
                <div className="w-full mb-2">
                  <InputField
                    id="edit-quantity"
                    type="number"
                    label="Quantity"
                    errorMessage={quantityErrorMessage}
                    onKeyDown={handleOnKeyDown}
                    {...field}
                  />
                </div>
              );
            }}
            name="quantity"
          />

          <Controller
            name="category"
            control={control}
            render={({ field: { value, onChange } }) => {
              const convertedValue = value.toString();
              return (
                <div className="w-full mb-2">
                  <SelectField
                    label="Category"
                    options={CATEGORY_PRODUCT}
                    value={convertedValue}
                    onChange={onChange}
                    className="py-2.5"
                  />
                </div>
              );
            }}
          />

          <Controller
            control={control}
            render={({ field }) => (
              <div className="w-full mb-2">
                <InputField
                  id="edit-provider"
                  label="Provider Name"
                  {...field}
                />
              </div>
            )}
            name="providerName"
          />
        </Flex>
      </Flex>
    </Card>
  );
};

export default ProductInfo;
