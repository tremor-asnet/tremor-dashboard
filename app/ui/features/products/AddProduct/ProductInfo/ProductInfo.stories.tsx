import type { Meta, StoryObj } from "@storybook/react";

// Libs
import { useForm } from "react-hook-form";

// Components
import ProductInfo from "./ProductInfo";

// Types
import { NewInfo } from "@/types";

const meta = {
  title: "Components/Products/AddProduct/ProductInfo",
  component: ProductInfo,
  tags: ["autodocs"],
  argTypes: {
    control: {
      description:
        "This is control from useForm of react-hook-form with type NewInfo({productName: string, description: string, providerName: string, weight: string, category: string, quantity: string})",
    },
  },
} as Meta<typeof ProductInfo>;

export default meta;

type Story = StoryObj<typeof meta>;

const ProductInfoFormProvider = () => {
  const { control } = useForm<NewInfo>({
    defaultValues: {
      productName: "Product A",
      description: "",
      weight: 0,
      category: 1,
      quantity: 0,
      providerName: "Provider A",
    },
  });

  return <ProductInfo control={control} />;
};

export const Default: Story = {
  render: () => <ProductInfoFormProvider />,
};
