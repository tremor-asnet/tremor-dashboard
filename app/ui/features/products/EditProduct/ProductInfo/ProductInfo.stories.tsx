// Libs
import { ReactNode, FC } from "react";
import { FormProvider, useForm } from "react-hook-form";
import type { Meta, StoryObj } from "@storybook/react";

// Components
import ProductInfo from "./ProductInfo";

const StorybookFormProvider: FC<{ children: ReactNode }> = ({ children }) => {
  const methods = useForm();
  return <FormProvider {...methods}>{children}</FormProvider>;
};

const meta = {
  title: "Components/Products/EditProduct/ProductInfo",
  component: ProductInfo,
  tags: ["autodocs"],
} as Meta<typeof ProductInfo>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  render: () => (
    <StorybookFormProvider>
      <ProductInfo />
    </StorybookFormProvider>
  ),
};
