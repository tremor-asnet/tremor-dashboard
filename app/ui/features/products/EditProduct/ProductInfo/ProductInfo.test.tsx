import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

// Components
import ProductInfo from "./ProductInfo";

// Constants
import { EXCEPT_KEYS } from "@/constants/common";

jest.mock("react-hook-form", () => {
  let mockReactHookForm = jest.requireActual("react-hook-form");

  mockReactHookForm.useFormContext = jest.fn();
  return {
    ...mockReactHookForm,
    useFormContext: jest.fn(() => {
      return {
        formState: {
          errors: {},
        },
      };
    }),
    Controller: (props: any) => {
      return (
        <>
          {props.render({
            field: {},
            formState: {
              errors: {},
            },
          })}
        </>
      );
    },
  };
});

describe("Testing ProductInfo component", () => {
  const handleOnKeyDown = jest.fn();

  it("should match snapshot", () => {
    const component = render(<ProductInfo />);
    expect(component).toMatchSnapshot();
  });

  it("should render title correctly", () => {
    const { getByText } = render(<ProductInfo />);
    expect(getByText("Product Information")).toBeInTheDocument();
  });

  it("renders pricing information fields correctly", () => {
    const { getByText, getByLabelText } = render(<ProductInfo />);

    expect(getByLabelText("Name")).toBeInTheDocument();
    expect(getByText("Description")).toBeInTheDocument();
    expect(getByLabelText("Weight")).toBeInTheDocument();
    expect(getByLabelText("Quantity")).toBeInTheDocument();
    expect(getByLabelText("Provider Name")).toBeInTheDocument();
  });

  it("should call preventDefault when key is in EXCEPT_KEYS.POSITIVE_DOUBLE", async () => {
    const mockEvent = {
      key: EXCEPT_KEYS.POSITIVE_DOUBLE[0],
      preventDefault: jest.fn(),
    };
    const { getByRole } = render(<ProductInfo />);
    const weightInput = getByRole("spinbutton", {
      name: /weight/i,
    });

    fireEvent.keyDown(weightInput, mockEvent); // Invalid input

    expect(weightInput).toBeInTheDocument();
  });

  it("should not call preventDefault when key is not in EXCEPT_KEYS.POSITIVE_DOUBLE", () => {
    const mockEvent = {
      key: "b",
      preventDefault: jest.fn(),
    };

    handleOnKeyDown(mockEvent);

    expect(mockEvent.preventDefault).not.toHaveBeenCalled();
  });
});
