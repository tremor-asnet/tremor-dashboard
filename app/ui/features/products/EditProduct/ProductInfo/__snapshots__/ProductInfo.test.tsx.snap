// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ProductInfo component should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="tremor-Card-root relative text-left rounded-tremor-default bg-tremor-background ring-tremor-ring shadow-tremor-card dark:ring-dark-tremor-ring dark:shadow-dark-tremor-card border-tremor-brand dark:border-dark-tremor-brand p-6 w-full dark:bg-dark-tremor-primary rounded-lg shadow-box-icon-default ring-0"
      >
        <p
          class="text-tremor-default text-primary dark:text-white font-bold text-xl mb-8"
        >
          Product Information
        </p>
        <div
          class="tremor-Flex-root flex w-full flex-col justify-between md:flex-row items-start gap-6"
        >
          <div
            class="tremor-Flex-root flex w-full flex-col justify-between items-center gap-4"
          >
            <div
              class="w-full mb-4"
            >
              <div
                class="w-full"
              >
                <label
                  class="text-secondary text-sm dark:text-lighter"
                  for="edit-name"
                >
                  Name
                </label>
                <input
                  class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
                  id="edit-name"
                  placeholder=""
                />
              </div>
            </div>
            <div
              class="w-full"
            >
              <p
                class="text-tremor-default text-secondary text-sm dark:text-lighter"
              >
                Description 
                <span
                  class="text-xs"
                >
                  (optional)
                </span>
              </p>
            </div>
          </div>
          <div
            class="tremor-Flex-root flex w-full flex-col justify-between items-center gap-6"
          >
            <div
              class="w-full mb-1"
            >
              <div
                class="w-full"
              >
                <label
                  class="text-secondary text-sm dark:text-lighter"
                  for="edit-weight"
                >
                  Weight
                </label>
                <input
                  class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
                  id="edit-weight"
                  placeholder=""
                  type="number"
                />
              </div>
            </div>
            <div
              class="w-full mb-2"
            >
              <div
                class="w-full"
              >
                <label
                  class="text-secondary text-sm dark:text-lighter"
                  for="edit-quantity"
                >
                  Quantity
                </label>
                <input
                  class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
                  id="edit-quantity"
                  placeholder=""
                  type="number"
                />
              </div>
            </div>
            <div
              class="w-full mb-2"
            >
              <div
                class="w-full"
              >
                <p
                  class="text-tremor-default text-secondary text-sm dark:text-lighter"
                >
                  Category
                </p>
                <select
                  class="peer h-full w-full appearance-none rounded-none border-b-2 border-gray-300 bg-transparent font-sans text-sm font-normal text-primary dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer outline outline-0 transition-all placeholder-shown:border-blue-gray-200 focus:border-gray-500 focus:outline-0 disabled:border-0 disabled:bg-blue-gray-50 placeholder:opacity-0 focus:placeholder:opacity-100 py-2.5"
                >
                  <option
                    class="dark:bg-primary"
                    data-testid="select-option"
                    value="1"
                  >
                    Clothing
                  </option>
                  <option
                    class="dark:bg-primary"
                    data-testid="select-option"
                    value="2"
                  >
                    Electronics
                  </option>
                  <option
                    class="dark:bg-primary"
                    data-testid="select-option"
                    value="3"
                  >
                    Furniture
                  </option>
                  <option
                    class="dark:bg-primary"
                    data-testid="select-option"
                    value="4"
                  >
                    Others
                  </option>
                  <option
                    class="dark:bg-primary"
                    data-testid="select-option"
                    value="5"
                  >
                    Real Estate
                  </option>
                </select>
              </div>
            </div>
            <div
              class="w-full mb-2"
            >
              <div
                class="w-full"
              >
                <label
                  class="text-secondary text-sm dark:text-lighter"
                  for="edit-provider"
                >
                  Provider Name
                </label>
                <input
                  class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
                  id="edit-provider"
                  placeholder=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="tremor-Card-root relative text-left rounded-tremor-default bg-tremor-background ring-tremor-ring shadow-tremor-card dark:ring-dark-tremor-ring dark:shadow-dark-tremor-card border-tremor-brand dark:border-dark-tremor-brand p-6 w-full dark:bg-dark-tremor-primary rounded-lg shadow-box-icon-default ring-0"
    >
      <p
        class="text-tremor-default text-primary dark:text-white font-bold text-xl mb-8"
      >
        Product Information
      </p>
      <div
        class="tremor-Flex-root flex w-full flex-col justify-between md:flex-row items-start gap-6"
      >
        <div
          class="tremor-Flex-root flex w-full flex-col justify-between items-center gap-4"
        >
          <div
            class="w-full mb-4"
          >
            <div
              class="w-full"
            >
              <label
                class="text-secondary text-sm dark:text-lighter"
                for="edit-name"
              >
                Name
              </label>
              <input
                class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
                id="edit-name"
                placeholder=""
              />
            </div>
          </div>
          <div
            class="w-full"
          >
            <p
              class="text-tremor-default text-secondary text-sm dark:text-lighter"
            >
              Description 
              <span
                class="text-xs"
              >
                (optional)
              </span>
            </p>
          </div>
        </div>
        <div
          class="tremor-Flex-root flex w-full flex-col justify-between items-center gap-6"
        >
          <div
            class="w-full mb-1"
          >
            <div
              class="w-full"
            >
              <label
                class="text-secondary text-sm dark:text-lighter"
                for="edit-weight"
              >
                Weight
              </label>
              <input
                class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
                id="edit-weight"
                placeholder=""
                type="number"
              />
            </div>
          </div>
          <div
            class="w-full mb-2"
          >
            <div
              class="w-full"
            >
              <label
                class="text-secondary text-sm dark:text-lighter"
                for="edit-quantity"
              >
                Quantity
              </label>
              <input
                class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
                id="edit-quantity"
                placeholder=""
                type="number"
              />
            </div>
          </div>
          <div
            class="w-full mb-2"
          >
            <div
              class="w-full"
            >
              <p
                class="text-tremor-default text-secondary text-sm dark:text-lighter"
              >
                Category
              </p>
              <select
                class="peer h-full w-full appearance-none rounded-none border-b-2 border-gray-300 bg-transparent font-sans text-sm font-normal text-primary dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer outline outline-0 transition-all placeholder-shown:border-blue-gray-200 focus:border-gray-500 focus:outline-0 disabled:border-0 disabled:bg-blue-gray-50 placeholder:opacity-0 focus:placeholder:opacity-100 py-2.5"
              >
                <option
                  class="dark:bg-primary"
                  data-testid="select-option"
                  value="1"
                >
                  Clothing
                </option>
                <option
                  class="dark:bg-primary"
                  data-testid="select-option"
                  value="2"
                >
                  Electronics
                </option>
                <option
                  class="dark:bg-primary"
                  data-testid="select-option"
                  value="3"
                >
                  Furniture
                </option>
                <option
                  class="dark:bg-primary"
                  data-testid="select-option"
                  value="4"
                >
                  Others
                </option>
                <option
                  class="dark:bg-primary"
                  data-testid="select-option"
                  value="5"
                >
                  Real Estate
                </option>
              </select>
            </div>
          </div>
          <div
            class="w-full mb-2"
          >
            <div
              class="w-full"
            >
              <label
                class="text-secondary text-sm dark:text-lighter"
                for="edit-provider"
              >
                Provider Name
              </label>
              <input
                class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
                id="edit-provider"
                placeholder=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
